helper/validation/map.go

```diff
@@ -0,0 +1,120 @@
+package validation
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/go-cty/cty"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// MapKeyLenBetween returns a SchemaValidateFunc which tests if the provided value
+// is of type map and the length of all keys are between min and max (inclusive)
+func MapKeyLenBetween(min, max int) schema.SchemaValidateDiagFunc {
+	return func(v interface{}, path cty.Path) diag.Diagnostics {
+		for key := range v.(map[string]interface{}) {
+			len := len(key)
+			if len < min || len > max {
+				return diag.Diagnostics{{
+					Severity:      diag.Error,
```

Comment:
As mentioned in my question answering it might be a better pattern to declare `var diags diag.Diagnostics` at the top and append to the slice, returning `diags` at the end. You could exit at the first error or actually accumulate all of them to return. Of course if situations where you cannot continue you should exit early, but in situations such as this were we can actually successfully collect multiple errors you have the ability to do so.

Reply:
Right now I'm exiting at the first error but it would make sense to accumulate a `Diagnostic` for every key (in this function) that is in error.
This will complicate the `Equals` helper or expanded equivalent as for example map iteration is non-deterministic and the `Diagnostics` entries will potentially be in different orders.
OK, I think I've just talked myself into sorting the keys and then iterating the map in a known order and accumulating all errors - The user's experience will be better in this case.